drafts/zip-daira-mempool-limitation.rst

```diff
@@ -0,0 +1,192 @@
+to be interpreted only as placing requirements on the zcashd implementation (and
+potentially other implementations that have adopted this specification in full).
+
+The mempool of a node holds a set of transactions. Each transaction has a cost,
+which is an integer defined as:
+
+  max(serialized transaction size in bytes, 4000) + fee_penalty
+
+where ``fee_penalty`` is 16000 if the transaction pays a fee less than
+10000 zatoshi, otherwise 0.
+
+Each node also MUST hold a FIFO queue RecentlyEvicted of pairs (txid, time), where
+the time indicates when the given txid was evicted. This SHOULD be empty on node
+startup. The size of RecentlyEvicted MUST never exceed 10000 entries.
+
+There MUST be a configuration option ``mempool.tx_cost_limit``, which SHOULD default
+to 80000000.
+
+There MUST be a configuration option ``mempool.eviction_memory_minutes``, which
+SHOULD default to 60.
+
+On receiving a transaction:
+
+* If it is in RecentlyEvicted, the transaction MUST be dropped.
+* Calculate its cost. If the total cost of transactions in the mempool including
+  this one would exceed ``mempool.tx_cost_limit``, then the node MUST repeatedly
+  call EvictTransaction (with the new transaction included as a candidate to evict)
+  until the total cost does not exceed ``mempool.tx_cost_limit``.
+
+EvictTransaction MUST do the following:
+
+* Select a random transaction to evict, weighted by cost.
+* Add the txid and the current time to RecentlyEvicted, dropping the oldest entry
+  in RecentlyEvicted if necessary to keep it to at most 10000 entries.
+* Remove it from the mempool.
+
+Nodes SHOULD remove transactions from RecentlyEvicted that were evicted more than
+``mempool.eviction_memory_minutes`` minutes ago. This MAY be done periodically,
+and/or just before RecentlyEvicted is accessed when receiving a transaction.
+
+
+Rationale
+=========
+
+The accounting for transaction size should include some overhead per transaction,
+to reflect the cost to the network of processing them (proof and signature
+verification; networking overheads; size of in-memory data structures). The
+implication of not including overhead is that a denial-of-service attacker would
+be likely to use minimum-size transactions so that more of them would fit in a
+block, increasing the unaccounted-for overhead. A possible counterargument would
+be that the complexity of accounting for this overhead is unwarranted given that
+the format of a transaction already imposes a minimum size. However, the proposed
+cost function is almost as simple as using transaction size directly.
+
+The threshold 4000 for the cost function is chosen so that the size in bytes of a
+typical fully shielded Sapling transaction (with, say, 2 shielded outputs and up
+to 5 shielded inputs) will fall below the threshold. This has the effect of
+ensuring that such transactions are not evicted preferentially to typical
+transparent transactions because of their size.
```

Comment:
We know of at least two legitimate use patterns that substantially diverge from this: 

1) Exchanges sweeping from user-specific ingress addresses to other parts of their fund management infrastructure will likely have a lot of inputs from consolidating small notes

2) Miner payouts will have lots of outputs

However, it seems reasonable to address the problem we have now based on the behavior we see now. Since this is not a consensus-critical ZIP it will be simple to address future problems that may arise.